src/lib/middleware/with-client-session.ts

```diff
@@ -13,71 +36,95 @@ export const withClientSession: Middleware<
     throw new UnauthorizedException({
-      type: "unauthorized",
-      message: "Client session tokens must start with seam_cst.",
+      type: "client_session_not_found",
+      message: "Unauthorized (client session not found)",
+    })
+  }
+
+  const {
+    client_session_id,
+    publishable_key,
+    api_key_id,
+    revoked_at,
+    expires_at,
+    workspace_id,
+    connected_account_ids,
+    connect_webview_ids,
+    user_identity_ids,
+  } = client_session
+
+  if (revoked_at != null && new Date(revoked_at) < new Date()) {
+    throw new UnauthorizedException({
+      type: "client_session_revoked",
+      message: "Unauthorized (client session revoked)",
     })
   }
 
-  if (short_token == null || long_token == null) {
-    throw new HttpException(400, {
-      type: "malformed_token",
-      message: "Malformed token",
+  if (new Date(expires_at) < new Date()) {
+    throw new UnauthorizedException({
+      type: "client_session_expired",
+      message: "Unauthorized (client session expired)",
     })
   }
 
-  if (is_cst) {
-    const client_session_token = req.db.client_sessions.find(
-      (cst) => cst.token === token,
-    )
+  const workspace = req.db.workspaces.find(
+    (w) => w.workspace_id === client_session.workspace_id,
+  )
+  if (workspace == null) {
+    throw new UnauthorizedException({
+      type: "workspace_not_found",
+      message: "Workspace was not found",
+    })
+  }
 
-    if (
-      client_session_token === null ||
-      typeof client_session_token === "undefined"
-    ) {
-      throw new UnauthorizedException({
-        type: "client_session_token_not_found",
-        message: "Client session token not found",
-      })
-    }
-    req.auth = {
-      type: "client_session",
-      workspace_id: client_session_token.workspace_id,
-      client_session_id: client_session_token.client_session_id,
-      connected_account_ids: client_session_token.connected_account_ids ?? [],
-      connect_webview_ids: client_session_token.connect_webview_ids ?? [],
-      publishable_key: null,
-      method: "api_key",
-    }
-    // Cannot run middleware after auth middleware.
-    // UPSTREAM: https://github.com/seamapi/nextlove/issues/118
-    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
-    return withSimulatedOutage(next as unknown as any)(
-      // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
-      req as unknown as any,
-      res,
-    )
+  if (publishable_key == null && api_key_id == null) {
+    throw new UnauthorizedException({
+      type: "unauthorized",
+      message: "publishable key or api key must be set",
+    })
   }
 
-  throw new HttpException(500, {
-    type: "unknown_auth_mode",
-    message: "Unknown Auth Mode",
-  })
+  req.auth = {
+    type: "client_session",
+    client_session_id,
+    method: api_key_id != null ? "api_key" : "publishable_key",
+    publishable_key: publishable_key ?? null,
+    api_key_id: api_key_id ?? null,
```

Comment:
Why does a client session have an api key id associated with it?

Reply:
Not sure, I was copying it from seam-connect ([source](https://github.com/seamapi/seam-connect/blob/75bedc4bf0a789ada4116ea14877014ec72784a5/lib/middlewares/with-client-session-token-auth.ts#L164)). Probably initially client session was mean to be created either using api key or publishable key?